lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java

```
@@ -389,13 +389,39 @@ private static void parseSegmentInfos(
     }
 
     long totalDocs = 0;
+    SegmentInfo info;
     for (int seg = 0; seg < numSegments; seg++) {
       String segName = input.readString();
       byte[] segmentID = new byte[StringHelper.ID_LENGTH];
       input.readBytes(segmentID, 0, segmentID.length);
       Codec codec = readCodec(input);
-      SegmentInfo info =
-          codec.segmentInfoFormat().read(directory, segName, segmentID, IOContext.READ);
+      try {
+        info = codec.segmentInfoFormat().read(directory, segName, segmentID, IOContext.READ);
+      } catch (FileNotFoundException | NoSuchFileException e) {
```

Hmm can we upgrade this to catch any "normal" (not crazy things like OOME) `Throwable`?  E.g. if the `.si` file had some bits scrambled or what not we will get a more exotic exception or maybe an exception because the checksum is mismatched.

Can we somehow return the root cause exception here and include it in `CheckIndexException` that we raise above?

---

Yes - will catch all exceptions and throw a new, specific `CorruptSegmentInfoException`

---

How about `IOException`? Tried catching all `Throwable t` here but it causes `TestIndexWrtier#testThreadInterruptDeadlock` to fail as it catches `ThreadInterruptExceptions` when it shouldn't

---

Exception handling is tricky!  I'm not sure what is the right exact set of exceptions to catch here ... but it should be broader than just the "no such file" exceptions I think.  Corruption in `.si` can result in exotic exceptions...